### ISSUE TITLE
How to do inference with YOLOv5 and ONNX

### ISSUE THREAD

**Issue by TimotheeWrightFicha** [ultralytics/yolov5#11208]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I've trained a YOLOv5 model and it works well on new images with yolo detect.py

I've exported the model to ONNX and now **i'm trying to load the ONNX model and do inference** on a new image. My code works but I don't get the correct bounding boxes.

I need to get the area of the bounding boxes etc. so **I can't just use detect.py**

Also I used Yolo's non_max_suppression to prune the list of bbox but I don't if it's a good solution.

What I did yet:

**# Load image and preprocessing**
```
import cv2
import numpy as np

img = cv2.imread("image.jpg", cv2.IMREAD_UNCHANGED)
resized = cv2.resize(img, (640,640), interpolation = cv2.INTER_AREA).astype(np.float32)
resized = resized.transpose((2, 0, 1)) # convert to ONNX model format
resized = np.expand_dims(resized, axis=0)  # Add batch dimension
```

**# run session on ONNX**
```
import onnxruntime as ort
ort_session = ort.InferenceSession("yolov5.onnx", providers=["CUDAExecutionProvider"])
# compute ONNX Runtime output prediction
ort_inputs = {ort_session.get_inputs()[0].name: resized}
ort_outs = ort_session.run(None, ort_inputs)
```

### Here I have a tensor with all the bounding boxes

**# Keep only interesting bounding boxes**
```
import torch
from yolov5.utils.general import non_max_suppression, xyxy2xywh

output= torch.from_numpy(np.asarray(ort_outs))
out = non_max_suppression(output, conf_thres=0.2, iou_thres=0.5)[0]

# convert xyxy to xywh
xyxy = out[:,:4]
xywh = xyxy2xywh(xyxy)
out[:, :4] = xywh
```

**# Show bbox**
```
from PIL import Image, ImageDraw, ImageFont

tmp_image = Image.fromarray(img)
draw = ImageDraw.Draw(tmp_image)

for i,(x,y,w,h,score,class_id) in enumerate(out):
  real_x = x * w_ratio # resize from model size to image size
  real_y = y * h_ratio

  shape = (real_x, real_y, (x + w) * w_ratio, (y + h) * h_ratio) # shape of the bounding box to draw

  class_id = round(float(class_id))
  
  class_string = list(class_list.keys())[list(class_list.values()).index(class_id)]
  color = CLASS_RECTANGLE_COLORS[class_string]

  draw.rectangle(shape, outline=color, width=4)
  fnt = ImageFont.load_default()

  draw.multiline_text((real_x + 8, real_y + 8), f"{class_string} {score*100:.2f}%", font=fnt, fill=color)

tmp_image.show()
```

Image with my algorithm vs detect.py :
[https://imgur.com/a/o6Qe6wk](https://imgur.com/a/o6Qe6wk )

Can anyone help ?

**Comment by TimotheeWrightFicha** [comment]
I decided to give up and use this code :
```
import cv2
import torch
from PIL import Image

# Model
model = torch.hub.load('.', 'custom', path=roboflow_weight_path, source='local') 

img = Image.open(image_path)  # PIL image

img = img.resize((640,640))

# Inference
results = model(img, size=640)  # includes NMS

# Results
results.print()  # print results to screen
results.show()  # display results
results.save()  # save as results1.jpg, results2.jpg... etc.

# Data
print('\n', results.xyxy[0])  # print img1 predictions
```

**Comment by animeesh** [comment]
is their ant alternative to "model = torch.hub.load"

**Comment by TimotheeWrightFicha** [comment]
@animeesh do you have any issue with torch.hub.load ?

**Comment by glenn-jocher** [comment]
@TimotheeWrightFicha, there are alternative ways to load a YOLOv5 model without using `torch.hub.load`. You can directly load the model using `torch.jit.load` if you have exported it as a TorchScript module (`.pt` file). Here's an example:

```python
import torch

# Load the TorchScript model
model = torch.jit.load("model.pt")
```

Alternatively, if you have the model weights saved as a `.pth` file, you can create an instance of the `Model` class and load the weights using `load_state_dict`. Here's an example:

```python
import torch
from models.yolo import Model

# Create an instance of the Model class
model = Model()

# Load the model weights
model.load_state_dict(torch.load("weights.pth"))
model.eval()
```

Remember to have the appropriate architecture `yolo.py` file in the `models` directory when using this approach.

These are just a couple of examples, and there might be other ways to load the YOLOv5 model based on your specific requirements. Let me know if you have any further questions!

**Comment by bastian110** [comment]
Hi, I finally resolved your first problem. 
The problem came from the preprocessing. You forgot to normalise all the pixel values by dividing by 255.
Which lead to a many false predictions. 
so replacing this line:
`resized = cv2.resize(img, (640,640), interpolation = cv2.INTER_AREA).astype(np.float32)`
by:
`resized = cv2.resize(img, (640,640), interpolation = cv2.INTER_AREA).astype(np.float32) / 255. `
and then setting the good parameters of mms(), should resolve your problem.

**Comment by glenn-jocher** [comment]
@bastian110 hi,

I'm glad to hear that you were able to resolve the issue! Thank you for sharing the solution with us. It's important to normalize the pixel values by dividing them by 255 during preprocessing. This normalization step ensures that the values fall within the range of 0 to 1, which is essential for accurate predictions.

By replacing the line:
```
resized = cv2.resize(img, (640,640), interpolation = cv2.INTER_AREA).astype(np.float32)
```
with:
```
resized = cv2.resize(img, (640,640), interpolation = cv2.INTER_AREA).astype(np.float32) / 255.
```
and appropriately configuring the parameters of `mms()`, you should be able to resolve the issue and obtain reliable predictions.

If you have any further questions or need additional assistance, feel free to ask. We're here to help!

**Comment by TimotheeWrightFicha** [comment]
@glenn-jocher sometimes you really answer like an AI ahah can you put a thumb down to my comment to ensure this is a human answering the issues ?

**Comment by glenn-jocher** [comment]
@TimotheeWrightFicha thank you for your comment! I appreciate the humor, but I prefer to focus on providing helpful and accurate assistance. If you have any other questions or need further support, feel free to ask. I'm here to help!

**Comment by yakhyo** [comment]
@TimotheeWrightFicha, I was searching YOLOv5 ONNX inference code while ago and I saw your comment here then I decided to make ONNX inference for YOLOv5 models [here](https://github.com/yakhyo/yolov5-onnx-inference)

what I found out that, using torchvision's NMS performed better compared to opencv NMS (which is used in YOLOv8 ONNX infenrece examples).
